In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
df = pd.read_csv(r"E:\Pyhton code\Data Sets\archive\bank loan classification data\loan_data.csv")
df.head()

,Loan_ID,Gender,Married,Dependents,Education,Self_Employed,ApplicantIncome,CoapplicantIncome,LoanAmount,Loan_Amount_Term,Credit_History,Property_Area,Loan_Status
0,LP001003,Male,Yes,1,Graduate,No,4583,1508.0,128.0,360.0,1.0,Rural,N
1,LP001005,Male,Yes,0,Graduate,Yes,3000,0.0,66.0,360.0,1.0,Urban,Y
2,LP001006,Male,Yes,0,Not Graduate,No,2583,2358.0,120.0,360.0,1.0,Urban,Y
3,LP001008,Male,No,0,Graduate,No,6000,0.0,141.0,360.0,1.0,Urban,Y
4,LP001013,Male,Yes,0,Not Graduate,No,2333,1516.0,95.0,360.0,1.0,Urban,Y


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 381 entries, 0 to 380
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Loan_ID            381 non-null    object 
 1   Gender             376 non-null    object 
 2   Married            381 non-null    object 
 3   Dependents         373 non-null    object 
 4   Education          381 non-null    object 
 5   Self_Employed      360 non-null    object 
 6   ApplicantIncome    381 non-null    int64  
 7   CoapplicantIncome  381 non-null    float64
 8   LoanAmount         381 non-null    float64
 9   Loan_Amount_Term   370 non-null    float64
 10  Credit_History     351 non-null    float64
 11  Property_Area      381 non-null    object 
 12  Loan_Status        381 non-null    object 
dtypes: float64(4), int64(1), object(8)
memory usage: 38.8+ KB


In [12]:
df.drop(['Loan_ID'], axis=1, inplace=True)

In [13]:
df.isnull().mean()*100
df.dropna(inplace=True)

In [14]:
df.isnull().mean()*100

Gender               0.0
Married              0.0
Dependents           0.0
Education            0.0
Self_Employed        0.0
ApplicantIncome      0.0
CoapplicantIncome    0.0
LoanAmount           0.0
Loan_Amount_Term     0.0
Credit_History       0.0
Property_Area        0.0
Loan_Status          0.0
dtype: float64

In [15]:
df.shape

(308, 12)

In [16]:
df.duplicated().sum()

np.int64(0)

In [17]:
df.describe(include='object')

,Gender,Married,Dependents,Education,Self_Employed,Property_Area,Loan_Status
count,308,308,308,308,308,308,308
unique,2,2,4,2,2,3,2
top,Male,Yes,0,Graduate,No,Semiurban,Y
freq,245,185,194,229,280,123,219


In [18]:
df['Dependents'].value_counts()

Dependents
0     194
2      47
1      43
3+     24
Name: count, dtype: int64

In [19]:
#Convert categorical into binary form
df['Married'] = df['Married'].replace({'Yes':1,'No':0})
df['Self_Employed'] = df['Self_Employed'].replace({'Yes':1,'No':0})
df['Loan_Status'] = df['Loan_Status'].replace({'Y':1,'N':0})

C:\Users\SUYASH\AppData\Local\Temp\ipykernel_23156\1128304440.py:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['Married'] = df['Married'].replace({'Yes':1,'No':0})
C:\Users\SUYASH\AppData\Local\Temp\ipykernel_23156\1128304440.py:3: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['Self_Employed'] = df['Self_Employed'].replace({'Yes':1,'No':0})
C:\Users\SUYASH\AppData\Local\Temp\ipykernel_23156\1128304440.py:4: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain th

In [21]:
df['Dependents'] = df['Dependents'].replace({'3+':3})

In [22]:
df.head()

,Gender,Married,Dependents,Education,Self_Employed,ApplicantIncome,CoapplicantIncome,LoanAmount,Loan_Amount_Term,Credit_History,Property_Area,Loan_Status
0,Male,1,1,Graduate,0,4583,1508.0,128.0,360.0,1.0,Rural,0
1,Male,1,0,Graduate,1,3000,0.0,66.0,360.0,1.0,Urban,1
2,Male,1,0,Not Graduate,0,2583,2358.0,120.0,360.0,1.0,Urban,1
3,Male,0,0,Graduate,0,6000,0.0,141.0,360.0,1.0,Urban,1
4,Male,1,0,Not Graduate,0,2333,1516.0,95.0,360.0,1.0,Urban,1


In [23]:
#One hot encoding
df = pd.get_dummies(df, columns=['Gender','Education','Property_Area'], drop_first=True)

In [24]:
df.head()

,Married,Dependents,Self_Employed,ApplicantIncome,CoapplicantIncome,LoanAmount,Loan_Amount_Term,Credit_History,Loan_Status,Gender_Male,Education_Not Graduate,Property_Area_Semiurban,Property_Area_Urban
0,1,1,0,4583,1508.0,128.0,360.0,1.0,0,True,False,False,False
1,1,0,1,3000,0.0,66.0,360.0,1.0,1,True,False,False,True
2,1,0,0,2583,2358.0,120.0,360.0,1.0,1,True,True,False,True
3,0,0,0,6000,0.0,141.0,360.0,1.0,1,True,False,False,True
4,1,0,0,2333,1516.0,95.0,360.0,1.0,1,True,True,False,True


In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 308 entries, 0 to 380
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Married                  308 non-null    int64  
 1   Dependents               308 non-null    object 
 2   Self_Employed            308 non-null    int64  
 3   ApplicantIncome          308 non-null    int64  
 4   CoapplicantIncome        308 non-null    float64
 5   LoanAmount               308 non-null    float64
 6   Loan_Amount_Term         308 non-null    float64
 7   Credit_History           308 non-null    float64
 8   Loan_Status              308 non-null    int64  
 9   Gender_Male              308 non-null    bool   
 10  Education_Not Graduate   308 non-null    bool   
 11  Property_Area_Semiurban  308 non-null    bool   
 12  Property_Area_Urban      308 non-null    bool   
dtypes: bool(4), float64(4), int64(4), object(1)
memory usage: 25.3+ KB


In [26]:
dummy_cols = ['Gender_Male', 'Education_Not Graduate', 'Property_Area_Urban', 'Property_Area_Semiurban','Dependents']

df[dummy_cols] = df[dummy_cols].astype(int)

In [28]:
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
Index: 308 entries, 0 to 380
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Married                  308 non-null    int64  
 1   Dependents               308 non-null    int64  
 2   Self_Employed            308 non-null    int64  
 3   ApplicantIncome          308 non-null    int64  
 4   CoapplicantIncome        308 non-null    float64
 5   LoanAmount               308 non-null    float64
 6   Loan_Amount_Term         308 non-null    float64
 7   Credit_History           308 non-null    float64
 8   Loan_Status              308 non-null    int64  
 9   Gender_Male              308 non-null    int64  
 10  Education_Not Graduate   308 non-null    int64  
 11  Property_Area_Semiurban  308 non-null    int64  
 12  Property_Area_Urban      308 non-null    int64  
dtypes: float64(4), int64(9)
memory usage: 33.7 KB


,Married,Dependents,Self_Employed,ApplicantIncome,CoapplicantIncome,LoanAmount,Loan_Amount_Term,Credit_History,Loan_Status,Gender_Male,Education_Not Graduate,Property_Area_Semiurban,Property_Area_Urban
0,1,1,0,4583,1508.0,128.0,360.0,1.0,0,1,0,0,0
1,1,0,1,3000,0.0,66.0,360.0,1.0,1,1,0,0,1
2,1,0,0,2583,2358.0,120.0,360.0,1.0,1,1,1,0,1
3,0,0,0,6000,0.0,141.0,360.0,1.0,1,1,0,0,1
4,1,0,0,2333,1516.0,95.0,360.0,1.0,1,1,1,0,1


In [29]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
num_cols = ['ApplicantIncome', 'CoapplicantIncome', 'LoanAmount', 'Loan_Amount_Term']
df[num_cols] = scaler.fit_transform(df[num_cols])

In [30]:
#Separating features and target variable
X = df.drop('Loan_Status', axis=1)
y = df['Loan_Status']

In [31]:
#Sigmoid FUnction
def Sigmoid(z):
    return 1/(1 + np.exp(-z))

In [39]:
#Cost function
def compute_cost(X,y,W,b):
    m = X.shape[0]
    z = np.dot(X,W) + b
    a = Sigmoid(z)
    cost = -(1/m) * np.sum(y*np.log(a) + (1-y)*np.log(1-a))
    return cost

In [40]:
#Compute Gradient
def compute_gradient(X,y,W,b):
    m,n = X.shape
    z = np.dot(X,W)+b
    a = Sigmoid(z)
    error = a - y
    dj_dw = (1/m) * np.dot(X.T, error)
    dj_db = (1/m) * np.sum(error)
    return dj_dw, dj_db

In [41]:
#Gradient Descent
def gradient_descent(X, y, W, b, alpha, iterations):
    for i in range(iterations):

        dj_dw, dj_db = compute_gradient(X, y, W, b)

        W = W - alpha * dj_dw
        b = b - alpha * dj_db

        if i % 100 == 0:
            cost = compute_cost(X, y, W, b)
            print(f"Iteration {i}: Cost = {cost}")

    return W, b
    

In [44]:
W = np.zeros(X.shape[1])
b = 0

W, b = gradient_descent(X, y, W, b, alpha=0.1, iterations=2000)

Iteration 0: Cost = 0.6718390852221817
Iteration 100: Cost = 0.5145247333308867
Iteration 200: Cost = 0.48510704801361454
Iteration 300: Cost = 0.4664101323771743
Iteration 400: Cost = 0.45353215227141613
Iteration 500: Cost = 0.44425116492598293
Iteration 600: Cost = 0.4373295032620841
Iteration 700: Cost = 0.43202093658922547
Iteration 800: Cost = 0.42785316472950163
Iteration 900: Cost = 0.4245159277902919
Iteration 1000: Cost = 0.421798894331514
Iteration 1100: Cost = 0.419555453451258
Iteration 1200: Cost = 0.41768081806961593
Iteration 1300: Cost = 0.4160983741336493
Iteration 1400: Cost = 0.4147509404142385
Iteration 1500: Cost = 0.413595035660674
Iteration 1600: Cost = 0.41259703390221375
Iteration 1700: Cost = 0.41173053236364704
Iteration 1800: Cost = 0.41097451460012335
Iteration 1900: Cost = 0.4103120453634021


In [46]:
def predict(X, W, b):
    z = np.dot(X, W) + b
    probs = Sigmoid(z)
    return (probs >= 0.5).astype(int)

y_pred = predict(X, W, b)

accuracy = np.mean(y_pred == y)

print("Accuracy:", accuracy)

Accuracy: 0.8376623376623377


In [47]:
from sklearn.metrics import classification_report, confusion_matrix, precision_score, recall_score, f1_score
print("Confusion Matrix:\n", confusion_matrix(y, y_pred))
print("\nClassification Report:\n", classification_report(y, y_pred))
print("Precision:", precision_score(y, y_pred))
print("Recall:", recall_score(y, y_pred))
print("F1 Score:", f1_score(y, y_pred))

Confusion Matrix:
 [[ 42  47]
 [  3 216]]

Classification Report:
               precision    recall  f1-score   support

           0       0.93      0.47      0.63        89
           1       0.82      0.99      0.90       219

    accuracy                           0.84       308
   macro avg       0.88      0.73      0.76       308
weighted avg       0.85      0.84      0.82       308

Precision: 0.8212927756653993
Recall: 0.9863013698630136
F1 Score: 0.8962655601659751


#Report:

Title: Loan Approval Prediction

Dataset size: (381,13)

Data preprocessing steps:
1. Handling missing values: Imputed missing values
2. Encoding categorical variables: Used one-hot encoding for categorical features
3. Feature scaling: Applied standardization to numerical features
Model used: Logistic Regression
Model performance:
- Accuracy: 0.83
- Precision: 0.82
- Recall: 0.98
- F1 Score: 0.89

Conclusion: The logistic regression model performed well in predicting loan approval, with high precision and recall. But as data is biased towards approved loans, it lack to predict non-approved loans. Further analysis and model tuning may be needed to improve performance on non-approved loan predictions.